docs/development/search.rst

```diff
@@ -0,0 +1,99 @@
+By default, Auto Indexing is turned off in development mode. To turn it on, change the
+`ELASTICSEARCH_DSL_AUTOSYNC` settings to `True` in the `readthedocs/settings/dev.py` file.
+After that, whenever a documentation successfully builds, or project gets added,
+the search index will update automatically.
+
+
+Architecture
+------------
+The search architecture is devided into 2 parts.
+One part is responsible for **indexing** the documents and projects and
+the other part is responsible for querying the Index to show the proper results to users.
+We use the `django-elasticsearch-dsl`_ package mostly to the keep the search working.
+`django-elasticsearch-dsl`_ is a wrapper around `elasticsearch-dsl`_ for easy configuration
+with Django.
+
+Indexing
+^^^^^^^^
+All the Sphinx documents are indexed into Elasticsearch after the build is successful.
+Currently, we do not index MkDocs documents to elasticsearch, but
+`any kind of help is welcome <https://github.com/rtfd/readthedocs.org/issues/1088>`_.
+
+How we index documentations
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+After any build is successfully finished, `HTMLFile` objects are created for each of the
+`HTML` files and the old version's `HTMLFile` object is deleted. The Signal_
+`bulk_post_create` is dispatched for created and `bulk_post_delete` is dispatched for deleted
+files. Both of the signals are dispatched with the list of the instances of `HTMLFile`
+in `instance_list` parameter.
+
+We listen to the `bulk_post_create` and `bulk_post_delete` signals in our `Search` application and
+index/delete the documentation content from the `HTMLFile` instances.
+
+
+How we index projects
+~~~~~~~~~~~~~~~~~~~~~
+We also index project information in our search index so that the user can search for projects
+from the main site. `django-elasticsearch-dsl`_ listen `post_create` and `post_delete` signals of
+`Project` model and index/delete into Elasticsearch accordingly.
+
+
+Elasticsearch Document
+~~~~~~~~~~~~~~~~~~~~~~
+
+`elasticsearch-dsl`_ provides model-like wrapper for the `Elasticsearch document`_.
+As per requirements of `django-elasticsearch-dsl`_, it is stored in the
+`readthedocs/search/documents.py` file.
+
+    **ProjectDocument:** It is used for indexing projects. Signal listener of
+    `django-elasticsearch-dsl`_ listens to the `post_save` signal of `Project` model and
+    then index/delete into Elasticsearch.
+
+    **PageDocument**: It is used for indexing documentation of projects. By default, the auto
+    indexing is turned off by `ignore_signals = settings.ES_PAGE_IGNORE_SIGNALS`.
+    `settings.ES_PAGE_IGNORE_SIGNALS` is `False` both in development and production.
+    As mentioned above, our `Search` app listens to the `bulk_post_create` and `bulk_post_delete`
+    signals and indexes/deleted documentation into Elasticsearch. The signal listeners are in
+    the `readthedocs/search/signals.py` file. Both of the signals are dispatched
+    after a successful documentation build.
+
+
```

Comment:
It's probably useful to mention how we parse the data from the JSON files to get the actual indexed data.